File: python/common/src/recoveryai_common/runtime/health.py
"""Readiness checks (plan Section 4.5)."""

import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from typing import Any

import asyncpg


@dataclass
class ReadinessCheck:
    name: str
    run: Callable[[], Awaitable[None]]
    optional: bool = False


@dataclass
class ReadinessRegistry:
    checks: list[ReadinessCheck] = field(default_factory=list)

    def register(self, check: ReadinessCheck) -> "ReadinessRegistry":
        self.checks.append(check)
        return self

    async def run(self) -> tuple[bool, list[dict[str, Any]]]:
        results: list[dict[str, Any]] = []
        ready = True
        for check in self.checks:
            started = time.perf_counter()
            try:
                await check.run()
                results.append(
                    {
                        "name": check.name,
                        "ok": True,
                        "durationMs": round((time.perf_counter() - started) * 1000),
                    }
                )
            except Exception as exc:
                results.append(
                    {
                        "name": check.name,
                        "ok": False,
                        "detail": str(exc),
                        "durationMs": round((time.perf_counter() - started) * 1000),
                    }
                )
                if not check.optional:
                    ready = False
        return ready, results


def postgres_readiness_check(database_url: str) -> ReadinessCheck:
    """A real round-trip to Postgres; a readiness probe that cannot fail is worthless."""

    async def run() -> None:
        connection = await asyncpg.connect(dsn=database_url, timeout=5)
        try:
            await connection.execute("select 1")
        finally:
            await connection.close()

    return ReadinessCheck(name="postgres", run=run)

File: packages/observability-ts/src/health.ts
import { SQL } from 'bun';

export interface ReadinessResult {
  name: string;
  ok: boolean;
  detail?: string;
  durationMs: number;
}

export interface ReadinessCheck {
  name: string;
  /** Optional checks report status but never make the service unready (plan Section 4.5). */
  optional?: boolean;
  run: () => Promise<void>;
}

/**
 * Readiness checks (plan Section 4.5). A service is ready when every required
 * dependency answers. An optional provider being disabled must not fail readiness.
 */
export class ReadinessRegistry {
  private readonly checks: ReadinessCheck[] = [];

  register(check: ReadinessCheck): this {
    this.checks.push(check);
    return this;
  }

  async run(): Promise<{ ready: boolean; checks: ReadinessResult[] }> {
    const results = await Promise.all(
      this.checks.map(async (check): Promise<ReadinessResult> => {
        const started = performance.now();
        try {
          await check.run();
          return {
            name: check.name,
            ok: true,
            durationMs: Math.round(performance.now() - started),
          };
        } catch (error) {
          return {
            name: check.name,
            ok: false,
            detail: error instanceof Error ? error.message : String(error),
            durationMs: Math.round(performance.now() - started),
          };
        }
      }),
    );

    const required = results.filter((r) => !this.checks.find((c) => c.name === r.name)?.optional);
    return { ready: required.every((r) => r.ok), checks: results };
  }
}

/**
 * Real Postgres connectivity check. Deliberately a genuine round-trip: a readiness
 * endpoint that always returns ok is worse than no readiness endpoint.
 */
export function postgresReadinessCheck(databaseUrl: string): ReadinessCheck {
  return {
    name: 'postgres',
    run: async () => {
      const sql = new SQL(databaseUrl, { max: 1, idleTimeout: 5 });
      try {
        await sql`select 1`;
      } finally {
        await sql.end();
      }
    },
  };
}
